admin/services/data-access/administration-message.data.service.js

```diff
@@ -0,0 +1,86 @@
+'use strict'
+
+const R = require('ramda')
+
+const sqlService = require('./sql.service')
+const TYPES = sqlService.TYPES
+const redisCacheService = require('../redis-cache.service')
+const administrationMessageDataService = {}
+
+/**
+ * Create a new service message
+ * @param data
+ * @return {Promise.<*>}
+ */
+administrationMessageDataService.sqlCreate = async (data) => {
+  await sqlService.create('serviceMessage', data)
+  return redisCacheService.drop('administrationMessageDataService.sqlFindServiceMessage')
+}
+
+/**
+ * Updates the service message
+ * @param data
+ * @return {Promise.<*>}
+ */
+administrationMessageDataService.sqlUpdate = async (data) => {
+  const params = [
+    {
+      name: 'title',
+      value: data.title,
+      type: TYPES.NVarChar
+    },
+    {
+      name: 'message',
+      value: data.message,
+      type: TYPES.NVarChar
+    },
+    {
+      name: 'updatedByUser_id',
+      value: data['updatedByUser_id'],
+      type: TYPES.Int
+    }
+  ]
+  const sql = `
+    UPDATE [mtc_admin].serviceMessage
+    SET title=@title, message=@message, updatedByUser_id=@updatedByUser_id
+    WHERE isDeleted=0`
+  await sqlService.modify(sql, params)
+  return redisCacheService.drop('administrationMessageDataService.sqlFindServiceMessage')
+}
+
+/**
+ * Set the service message as deleted.
+ * @returns {Promise.<void>}
+ */
+administrationMessageDataService.sqlDeleteServiceMessage = async (data) => {
+  const params = [
+    {
+      name: 'deletedByUser_id',
+      value: data['deletedByUser_id'],
+      type: TYPES.Int
+    }
+  ]
+  const sql = `
+    UPDATE [mtc_admin].serviceMessage
+    SET isDeleted=1, deletedByUser_id=@deletedByUser_id
+    WHERE isDeleted=0
+  `
+  await sqlService.modify(sql, params)
+  return redisCacheService.drop('administrationMessageDataService.sqlFindServiceMessage')
+}
+
+/**
+ * Fetch active service message
+ * @return {Promise<Object>}
+ */
+administrationMessageDataService.sqlFindActiveServiceMessage = async () => {
+  const sql = `
+    SELECT TOP 1 *
+    FROM [mtc_admin].serviceMessage
+    WHERE isDeleted = 0
+  `
+  const result = await sqlService.query(sql, [], 'administrationMessageDataService.sqlFindServiceMessage')
+  return R.head(result)
```

Comment:
so we are always getting it from SQL? you implement dropping from redis, but you never populate redis

Reply:
Line 82: has an extra parameter for the method that we wish to cache.
In this case it's `administrationMessageDataService.sqlFindServiceMessage`
The caching is happening within sqlService which is exactly the same pattern that check window utilises.
At school home level they are expected to query using this particular method which is cached in redis.
As we discussed it's not really nice that redis is implemented as a dependency on the sql service level.

Reply:
as discussed, please move redis to admin message service, and reinforce with unit tests for cache invalidation, fetching etc

Reply:
@GuyHarwood updated